OIDC Workshop Blog Post (#1407)
oktadeveloper/okta-blog: _source/_posts/2023-07-28-oidc_workshop.md

The frontend will hit the backend's `/openid/start/${org_id}` endpoint to initiate the OpenID login flow when a user belongs to an org. 

When that endpoint is hit, it will identify the org and create an appropriate strategy for it. It will then handle any errors that might result, and call `passport.authenticate` to use the OpenID configuration passed from the database through the newly created `Strategy` and authenticate the user. 

```
// The frontend then redirects here to have the backend start the OIDC flow.
// (You should probably use random IDs, not auto-increment integers
// to avoid revealing how many enterprise customers you have.)
app.get('/openid/start/:id', async (req, res, next) => {

  const org = await orgFromId(req.params.id);
  if(!org) {
    return res.sendStatus(404);
  }

  const strategy = createStrategy(org);
  if(!strategy) {
    return res.sendStatus(404);
  }

  passport.authenticate(strategy)(req, res, next);

});
```

For passport to authenticate the user, it will redirect them to their identity provider so that they can log in. After the OpenID server authenticates the user, it redirects them back to the application's `callback` url. 

> To recap, a call to the backend's `/openid/start/` endpoint with an org's ID will use passport to redirect the user to their org's OpenID server. The user will prove their identity to their org's OpenID server, and that server will return the authenticated session to the Todo app via passport. Only the app's backend will touch the session information from the OIDC server, so frontend tampering cannot intercept the user's session. 

Author: please fact-check here; I made this part up

---
layout: blog_post

> title: "Enterprise Readiness Workshop: OpenID Connect"

Author: ```suggestion
title: "Enterprise-Ready Workshop: OpenID Connect"
```

Today, we'll walk through adding OIDC to our Todo sample application. 

For an in-depth walkthrough of developing and troubleshooting the code in this workshop, follow along on with the video: 
{% youtube -gwgEAa4TkU %}

{% include toc.md %}

## Before You Begin: 

Follow the [getting started guide](/blog/2023/07/27/enterprise-ready-getting-started) to set up the Todo app and its dependencies. Make sure you can run the app locally and view its login page in your browser. Launch Prisma studio to examine the database, as well. 

You will also need a free [Okta Developer Account](https://developer.okta.com/login/), to test the setup steps your customer will follow when integrating their OpenID Connect server with your application. 

## The Problem

The Todo app currently only supports password login, but your enterprise customers' IT departments don't want their users managing one-off passwords. You would also like to use customer data, such as which user accounts belong to which organizations, to enhance your app's features. 

> Since you value flexibility and maintainability, you want to use a passport library to let your app integrate with arbitrary OpenID Connect servers. 

Reviewer: ```suggestion
Since you value flexibility and maintainability, you want to use a library to let your app integrate with arbitrary OpenID Connect servers. 
```

## Add Organizations

To support OpenID Connect, the app needs to know about more than just users and Todo items. It needs a new model, called the `org`, which represents an organization that users can belong to. 

The `org`'s ID will be an auto-increment field for demonstration purposes, although using sequential identifiers in production could leak information about how many customers you have.

The `org` is identified by its email domain, which is a `string` field. 

Since your app will have many customers with different identity providers, you'll store each org's OpenID configuration in the database. This will require columns for: 

- Authorization endpoint, a URL where the app will redirect users from the org so they can log in
- Token endpoint, location on the OIDC server where the app can get an ID token
- User info endpoint, where the app can retrieve the user's profile info like name and email
- Client ID and client secret, which are the credentials that the app can use to authenticate to the org's OIDC server
- API Key, for future use, if you add API access to the app for users in this org
- Issuer, an identifier representing the OIDC server for the org

> 

Reviewer: ```suggestion

Note that to keep things simple for demonstration purposes, we are storing the client secret and API key in plain text in the database. In production, you should use some sort of encrypted key store to manage these credentials.
```

Here's how organization support looks in `Schema.prisma`:
```
model Org {
  id        Int     @id @default(autoincrement())
  domain    String  @unique
  issuer                 String @default("")
  authorization_endpoint String @default("")
  token_endpoint         String @default("")
  userinfo_endpoint      String @default("")
  client_id              String @default("")
  client_secret          String @default("")
  apikey                 String
  Todo     Todo[]
  User     User[]
}
```

## Enhance Users

The sample application already manages rudimentary user accounts, but with OIDC support, users will belong to organizations. The user model in `schema.prisma` needs a relationship to an organization, which is implemented with the `org` and `orgID` fields. 

> Previously, the sample app treated a user's email address as their unique identifier. This can cause problems if the user's organization changes its domain name, or if the user's email address changes when they change their name. Instead, with OIDC support, a unique identifier for each user can be supplied by their identity provider. This identifier is called `subject` in the OIDC protocol, and is guaranteed to be a unique and stable identifier for that user on that server. It's possible for multiple OIDC servers to issue the same `subject` ID, but the combination of server and subject uniquely identifies a single user.

Reviewer: ```suggestion
Previously, the sample app treated a user's email address as their unique identifier. This can cause problems if the user's organization changes its domain name, or if the user's email address changes when they change their name. Instead, with OIDC support, a unique identifier for each user can be supplied by their identity provider. This identifier is called `sub` (short for "subject") in the OIDC protocol, and is guaranteed to be a unique and stable identifier for that user on that server. It's possible for multiple OIDC servers to issue the same `sub` ID, since this is not a globally unique value, so the combination of server and subject uniquely identifies a single user.
```

  userId    Int
  org       Org?    @relation(fields: [orgId], references: [id])
  orgId     Int?
}
```

## Migrate Database

After saving these changes to `schema.prisma`, open a terminal and run `npx prisma generate` to generate the new schema, and `npx prisma db push` to apply it. 

After completing this migration, you can explore the new database structure in Prisma. Users belong to organizations, and tasks can be associated with orgs. 

# Frontend Changes

If a user is able to log in via OpenID Connect, what changes will the frontend need around the login experience? Right now, every user is prompted for their username and password at the login page. But OIDC users won't have a password for this app, and should not entrust their login credentials to your web form! 

To handle both OIDC and non-OIDC users, the app should first prompt for the user's email address, and check whether they should be redirected to their organization's identity provider. If a user doesn't belong to an org, they should instead enter their password to log in. You can now modify the sample app's frontend code to support this behavior. 

## Hide password field

> The first change in `src/app/components/signin.tsx` is hiding the password field at first. Find the `div` containing the password, give it an identifier, and set it to be hidden by default: 

Reviewer: ```suggestion
The first change in `src/app/components/signin.tsx` is hiding the password field when the page is first loaded. Find the `div` containing the password, give it an identifier, and set it to be hidden by default: 
```

```
Implement `onUsernameEnteredFn` by modifying a copy of `onAuthenticateFn` in `authState.tsx`. Instead of using the `/api/signin` endpoint on the app's backend, you'll use the `/api/openid/check` endpoint:  

```
  const onUsernameEnteredFn = async (username: string) => {
    const url = `/api/openid/check`;
    try {
        const res = await fetch(url, {
          method: 'POST',
          headers: {
            'Content-Type': 'application/json'
          },
          body: JSON.stringify({ username })
        });

        const {org_id} = await res.json();

        return org_id;
      } catch (error: unknown) {
        console.error(error);
      }

    return null;
  }
```

And finally, make sure that `onUsernameEnteredFn` is included in the provider context returned from `onRevokeAuthFn`: 

```
  return <AuthContext.Provider value={{ authState, onAuthenticateFn, onUsernameEnteredFn, onRevokeAuthFn, userIsAuthenticatedFn }}>{children}</AuthContext.Provider>;
```

From the perspective of the app's frontend, these are all the changes that matter. Supporting OIDC will not change how the frontend communicates with the backend through sessions. 

> The frontend is now ready to use the backend's `/api/openid/check` endpoint to investigate whether a user belongs to an OIDC org, and then hit the org's custom `/api/openid/start/${org_id}` endpoint to initiate the login flow if an org is found! Your next step is to add these endpoints to the application's backend. 

Reviewer: ```suggestion
The frontend is now ready to use the backend's `/api/openid/check` endpoint to investigate whether a user belongs to an OIDC org, and then redirect the browser to the org's custom `/api/openid/start/${org_id}` endpoint to initiate the login flow if an org is found! Your next step is to add these endpoints to the application's backend. 
```

# Backend Changes

The application's backend will use a library to handle most of the OIDC authentication process. After authenticating, the user will be redirected back to the app just as they are when logging in with a password.

## Install Passport OIDC Library

> Since the sample app is using Passport for authentication, the Passport OIDC library will offer seamless integration. 

Reviewer: ```suggestion
Since the sample app is already using Passport for password authentication, the Passport OIDC library will offer seamless integration with the OpenID Connect provider.
```

In your project, run `npm install --save passport-openidconnect` to install the library.  This will update `package.json` automatically to specify the installed version: 

```
    "passport-openidconnect": "^0.1.1",
```

## Add Helper Functions

Add the following helper functions to `apps/api/src/main.ts` to simplify your upcoming work. 

> `orgFromId` will take an integer org ID and return the first database result for an organization that the ID references:

Reviewer: ```suggestion
`orgFromId` will take an integer org ID and return the database record for the organization that the ID references:
```

```
async function orgFromId(id) {
  const org = await prisma.org.findFirst({
    where: {
      id: parseInt(id)
    }
  })
  return org
}
```
`getDomainFromEmail` will parse a domain from an email address, by returning everything after the `@`: 

```
function getDomainFromEmail(email) {
  let domain;
  try {
    domain = email.split('@')[1];
  } catch(e) {
    return null;
  }
  return domain;
}
```

## Add OpenID Org Check Route

> The `api/openid/check` endpoint will return the numeric ID of the org that a user's email domain belongs to, or `null` if the user isn't part of an org. The user's full email address will arrive at the endpoint in the request body, so the helper function `getDomainFromEmail` will return the domain, before it's used to look up the corresponsding org in the database. Add this code to `main.ts`: 

Reviewer: ```suggestion
The `/api/openid/check` endpoint will return the numeric ID of the org that a user's email domain belongs to, or `null` if the user isn't part of an org. The user's full email address will arrive at the endpoint in the request body, so the helper function `getDomainFromEmail` will return the domain, before it's used to look up the corresponsding org in the database. Add this code to `main.ts`: 
```

```
app.post('/api/openid/check', async (req, res, next) => {
  const { username } = req.body;

  const domain = getDomainFromEmail(username);
  if(domain) {
    var org = await prisma.org.findFirst({
      where: {
        domain: domain
      }
    });
    if(!org) {
      org = await prisma.org.findFirst({
        where: {
          User: {
            some: {
              email: username
            }
          }
        }
      })
    }
    if(org && org.issuer) {
      return res.json({ org_id: org.id });
    }
  }

  res.json({ org_id: null });
});
```

> If the user belongs to an org and the database knows the `issuer`, or OIDC server, for that org, the org's ID will be returned. Otherwise, a `null` result will indicate that the user is not part of an org that the app knows about. 

Reviewer: ```suggestion
If the user belongs to an org and the database knows the `issuer` (the OIDC server) for that org, the org's ID will be returned. Otherwise, a `null` result will indicate that the user is not part of an org that the app knows about. 
```

    var user = await prisma.user.findFirst({
      where: {
        orgId: org.id,
        externalId: profile.id,
      }
    })

    if(!user) {
      user = await prisma.user.findFirst({
        where: {
          orgId: org.id,
          email: profile.emails[0].value,
        }
      })
      if(user) {
        await prisma.user.update({
          where: {id: user.id},
          data: {externalId: profile.id}
        })
      }
    }

    if(!user) {
      user = await prisma.user.create({
        data: {
          org: {connect: {id: org.id}},
          externalId: profile.id,
          email: profile.emails[0].value,
          name: profile.displayName,
        }
      })
    }

    return cb(null, user);
  })
}
```
What path will this code follow if the user existed in the database but needed to be updated? What path will it follow if this is the user's first time logging into the Todo app?

Now that the Passport OIDC library is configured to interface with the database and update it as necessary whenever a user logs in, it's time to implement the endpoints which will initiate OIDC auth by calling the `passport.authenticate` function. 

## Start the OpenID Flow

> The frontend will hit the backend's `/openid/start/${org_id}` endpoint to initiate the OpenID login flow when a user belongs to an org. 

Reviewer: ```suggestion
The frontend will redirect the browser to the backend's `/openid/start/${org_id}` endpoint to initiate the OpenID login flow when a user belongs to an org. 
```

When that endpoint is hit, it will identify the org and create an appropriate strategy for it. It will then handle any errors that might result, and call `passport.authenticate` to use the OpenID configuration passed from the database through the newly created `Strategy` and authenticate the user. 

```
// The frontend then redirects here to have the backend start the OIDC flow.
// (You should probably use random IDs, not auto-increment integers
// to avoid revealing how many enterprise customers you have.)
app.get('/openid/start/:id', async (req, res, next) => {

  const org = await orgFromId(req.params.id);
  if(!org) {
    return res.sendStatus(404);
  }

  const strategy = createStrategy(org);
  if(!strategy) {
    return res.sendStatus(404);
  }

  passport.authenticate(strategy)(req, res, next);

});
```

For passport to authenticate the user, it will redirect them to their identity provider so that they can log in. After the OpenID server authenticates the user, it redirects them back to the application's `callback` url. 

> To recap, a call to the backend's `/openid/start/` endpoint with an org's ID will use passport to redirect the user to their org's OpenID server. The user will prove their identity to their org's OpenID server, and that server will return the authenticated session to the Todo app via passport. Only the app's backend will touch the session information from the OIDC server, so frontend tampering cannot intercept the user's session. 

Reviewer: ```suggestion
To recap, redirecting the user to the backend's `/openid/start/` endpoint with an org's ID will use passport to redirect the user to their org's OpenID server. The user will prove their identity to their org's OpenID server, and that server will redirect the user to the app's backend, at which point Passport will start a session to the Todo app. Only the app's backend will see the information from the OIDC server, so frontend tampering cannot intercept the user's session. 
```

app.get('/openid/callback/:id', async (req, res, next) => {

  const org = await orgFromId(req.params.id);
  if(!org) {
    return res.sendStatus(404);
  }

  const strategy = createStrategy(org);
  if(!strategy) {
    return res.sendStatus(404);
  }

  passport.authenticate(strategy, {
    successRedirect: 'http://localhost:3000/'
  })(req, res, next);

});
```

What kinds of problem is this code anticipating? Where will the user be sent if their authentication is successful? 

# Connect to Identity Provider

Now the app is ready for customers to connect their identity providers! This workshop will use an Okta Developer Account as the OIDC provider, to simulate a customer using Okta to connect with your app. 

In a production setting, you would have your customers provide you with information about their identity providers. But in this workshop, you're pretending to be your own customer, so you'll gather the OIDC provider information in this step.

Visit [developer.okta.com](developer.okta.com) and log in to your Developer Account, or sign up if you don't have an account yet. Open the admin console if you're redirected to your user account dashboard. 

## Create App Integration

> In the Okta admin console, navigate to "Applications" under the "Applications" heading in the left sidebar. Click the "Create App Integration" button, because your sample app isn't published to the app catalog. (If your app's audience includes a lot of Okta customers, publishing to the Okta catalog can simplify their onboarding process.) 

Reviewer: ```suggestion
In the Okta admin console, navigate to "Applications" under the "Applications" heading in the left sidebar. Click the "Create App Integration" button, because your sample app isn't published to the app catalog. (If your app's audience includes a lot of Okta customers, publishing to the Okta Integration Network can simplify their onboarding process.) 
```

---
layout: blog_post
title: "Enterprise-Ready Workshop: OpenID Connect"

> author: [edunham, aaron-parecki]

Author: ```suggestion
author: [aaron-parecki, edunham]
```
